zipline/pipeline/loaders/events.py

```diff
@@ -0,0 +1,203 @@
+        #  purposes.
+        self.events_by_sid = (
+            events_by_sid.copy()
+        )
+        dates = self.all_dates.values
+
+        for k, v in iteritems(events_by_sid):
+            # Already a DataFrame
+            if isinstance(v, pd.DataFrame):
+                if TS_FIELD_NAME not in v.columns:
+                    if not infer_timestamps:
+                        raise ValueError(
+                            DF_NO_TS_NOT_INFER_TS_ERROR.format(
+                                timestamp_column_name=TS_FIELD_NAME,
+                                sid=k
+                            )
+                        )
+                    self.events_by_sid[k] = v = v.copy()
+                    v.index = [dates[0]] * len(v)
+                else:
+                    self.events_by_sid[k] = v.set_index(TS_FIELD_NAME)
+                # Once data is in a DF, make sure columns are correct.
+                cols_except_ts = (set(v.columns) -
+                                  {TS_FIELD_NAME})
+
+                # Check that all columns other than timestamp are as expected.
+                if cols_except_ts != self.expected_cols:
+                    raise ValueError(
+                        WRONG_COLS_ERROR .format(
+                            expected_columns=self.expected_cols,
+                            sid=k,
+                            resulting_columns=v.columns.values
+                        )
+                    )
+            # Not a DataFrame and we only expect 1 column
+            elif len(self.expected_cols) == 1:
+                # First, must convert to DataFrame.
+                if isinstance(v, pd.Series):
+                    if not isinstance(v.index, pd.DatetimeIndex):
+                        raise ValueError(
+                            SERIES_NO_DTINDEX_ERROR.format(sid=k)
+                        )
+                    self.events_by_sid[k] = pd.DataFrame({
+                        list(self.expected_cols)[0]: v})
+                elif isinstance(v, pd.DatetimeIndex):
+                    if not infer_timestamps:
+                        raise ValueError(
+                            DTINDEX_NOT_INFER_TS_ERROR.format(sid=k)
+                        )
+                    self.events_by_sid[k] = pd.DataFrame({
+                        list(self.expected_cols)[0]: v
+                    }, index=[dates[0]] * len(v))
+                else:
+                    # We expect 1 column, but we got something other than a
+                    # Series, DatetimeIndex, or DataFrame.
+                    raise ValueError(WRONG_SINGLE_COL_DATA_FORMAT_ERROR
+                                     .format(sid=k))
+            else:
+                # We expected multiple columns, but we got something other
+                # than a DataFrame.
+                raise ValueError(WRONG_MANY_COL_DATA_FORMAT_ERROR
+                                 .format(sid=k))
```

Comment:
Could we try to not break the line on an attribute access. I find this style really hard to read